qiskit/pulse/instructions/play.py

```diff
@@ -55,7 +62,7 @@ def __init__(self, pulse: Pulse, channel: PulseChannel, name: Optional[str] = No
         super().__init__(operands=(pulse, channel), name=name)
 
     @property
-    def pulse(self) -> Pulse:
+    def pulse(self) -> [Pulse, Signal]:
```

Comment:
I think we should internally convert pulse into signal by taking `Signal(Pulse, None)` if a frame is not provided. In this way we can avoid inserting branching logic in following transforms. We can add `Play.signal` property rather than returning `Union[Pulse, Signal]` here. This property should return only `self.signal.pulse`, otherwise user always need to check instance to get pulse envelope.

Reply:
Sounds good. See here: d51d2d0